File: 7.maze-solver/gprimitives.py
from time import sleep

class Point():
  def __init__(self, x, y):
    self.x = x
    self.y = y

class Line():
  def __init__(self, start: Point, end: Point):
    self.start = start
    self.end = end
  
  def draw(self, canvas, fill_color):
    canvas.create_line(self.start.x, self.start.y, self.end.x, self.end.y, fill=fill_color, width = 2)
    canvas.pack()


class Cell():
  def __init__(self, top_left: Point, bottom_right: Point, window):
    self.__x1 = top_left.x
    self.__y1 = top_left.y
    self.__x2 = bottom_right.x
    self.__y2 = bottom_right.y
    self.has_left_wall = True
    self.has_right_wall = True
    self.has_top_wall = True
    self.has_bottom_wall = True
    self.__win = window
    self.visited = False
  
  def draw(self):

    bg_color = self.__win.background
    left_line = Line(Point(self.__x1, self.__y1), Point(self.__x1, self.__y2))
    if self.has_left_wall:
      self.__win.draw_line(left_line)
    else:
      self.__win.draw_line(left_line, bg_color)

    right_line = Line(Point(self.__x2, self.__y1), Point(self.__x2, self.__y2))
    if self.has_right_wall:
      self.__win.draw_line(right_line)
    else:
      self.__win.draw_line(right_line, bg_color)
    
    top_line = Line(Point(self.__x1, self.__y1), Point(self.__x2, self.__y1))
    if self.has_top_wall:
      self.__win.draw_line(top_line)
    else:
      self.__win.draw_line(top_line, bg_color)
    
    bottom_line = Line(Point(self.__x1, self.__y2), Point(self.__x2, self.__y2))
    if self.has_bottom_wall:
      self.__win.draw_line(bottom_line)
    else:
      self.__win.draw_line(bottom_line, bg_color)
  
  def draw_move(self, to_cell, undo = False):
    Point1 = Point((self.__x1 + self.__x2) / 2, (self.__y1 + self.__y2) / 2)
    Point2 = Point((to_cell.__x1 + to_cell.__x2) / 2, (to_cell.__y1 + to_cell.__y2) / 2)
    if undo:
      self.__win.draw_line(Line(Point1, Point2), "#aaa")
      sleep(0.0005)
    else:
      self.__win.draw_line(Line(Point1, Point2), "red")

  def _draw_emoji(self, emoji, font_size = 40):

    x = (self.__x1 + self.__x2) / 2
    y = (self.__y1 + self.__y2) / 2
    self.__win.get_canvas().create_text(x, y, text=emoji,font=('Arial', 40), anchor="center")

File: 7.maze-solver-html5-js/gprimitives.js
class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
}

class Line {
  constructor(start, end) {
    this.start = start;
    this.end = end;
  }

  draw(ctx, fillColor = 'black', lineWidth = 2) {
    ctx.beginPath();
    ctx.moveTo(this.start.x, this.start.y);
    ctx.lineTo(this.end.x, this.end.y);
    ctx.strokeStyle = fillColor;
    ctx.lineWidth = lineWidth;
    ctx.stroke();
  }
}

class Cell {
  constructor(topLeft, bottomRight, canvas) {
    this._x1 = topLeft.x;
    this._y1 = topLeft.y;
    this._x2 = bottomRight.x;
    this._y2 = bottomRight.y;
    this.hasLeftWall = true;
    this.hasRightWall = true;
    this.hasTopWall = true;
    this.hasBottomWall = true;
    this.canvas = canvas;
    this.ctx = this.canvas.getContext('2d');
    this.visited = false;
  }

  draw() {
    const bgColor = "#ccc"; //window.getComputedStyle(this.canvas).backgroundColor;

    const drawLine = (line, color = 'black') => line.draw(this.ctx, color);
    const eraseLine = (line, color = bgColor) => line.draw(this.ctx, color, 3);

    const leftLine = new Line(new Point(this._x1, this._y1), new Point(this._x1, this._y2));
    this.hasLeftWall ? drawLine(leftLine) : eraseLine(leftLine, bgColor);

    const rightLine = new Line(new Point(this._x2, this._y1), new Point(this._x2, this._y2));
    this.hasRightWall ? drawLine(rightLine) : eraseLine(rightLine, bgColor);

    const topLine = new Line(new Point(this._x1, this._y1), new Point(this._x2, this._y1));
    this.hasTopWall ? drawLine(topLine) : eraseLine(topLine, bgColor);

    const bottomLine = new Line(new Point(this._x1, this._y2), new Point(this._x2, this._y2));
    this.hasBottomWall ? drawLine(bottomLine) : eraseLine(bottomLine, bgColor);
  }

  drawMove(toCell, undo = false) {
    const midPoint = (a, b) => (a + b) / 2;
    const point1 = new Point(midPoint(this._x1, this._x2), midPoint(this._y1, this._y2));
    const point2 = new Point(midPoint(toCell._x1, toCell._x2), midPoint(toCell._y1, toCell._y2));
    const line = new Line(point1, point2);
    if (undo) {
      line.draw(this.ctx, "#aaa");
      setTimeout(() => {}, 100); // Simulate a delay
    } else {
      line.draw(this.ctx, "red", 3);
    }
  }

  min(a, b) {
    return a < b ? a : b;
  }

  drawEmoji(emoji, fontSize = 40) {
    const x = (this._x1 + this._x2) / 2;
    const y = (this._y1 + this._y2) / 2;

    this.ctx.font = `${this.min(this.min(fontSize, Math.abs(this._x2 - this._x1 -6)), Math.abs(this._y2 - this._y1 -6))}px "Noto Color Emoji", "Apple Color Emoji", "Segoe UI Emoji", sans-serif`;
    this.ctx.textAlign = "center"; 
    this.ctx.textBaseline = "middle"; 

    this.ctx.fillText(emoji, x, y);
  }
  
}

// Example usage
// document.addEventListener('DOMContentLoaded', () => {
//   const canvas = document.getElementById('myCanvas');
  
//   if (!canvas) {
//     console.error("Canvas element not found");
//   }
// });
